lib/src/api_analysis/summary.dart

```diff
@@ -0,0 +1,351 @@
+// Copyright (c) 2022, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/analysis/results.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:collection/collection.dart';
+import 'package:pana/pana.dart';
+import 'package:path/path.dart' as path;
+
+import 'common.dart';
+import 'normalize_shape.dart';
+import 'shapes.dart';
+import 'shapes_ext.dart';
+
+Future<PackageShape> summarizePackage({
+  required PackageAnalysisContext context,
+  required String packagePath,
+  required bool normalize,
+}) async {
+  final pubspecString = context.readFile(path.join(
+    packagePath,
+    'pubspec.yaml',
+  ));
+  final pubspec = Pubspec.parseYaml(pubspecString);
+
+  if (pubspec.version == null) {
+    throw StateError(
+        'Specify a version number in `pubspec.yaml` before running analysis.');
+  }
+
+  final package = PackageShape(
+    name: pubspec.name,
+    version: pubspec.version!.toString(),
+    libraries: <LibraryShape>[],
+    getters: <GlobalPropertyShape>[],
+    setters: <GlobalPropertyShape>[],
+    functions: <FunctionShape>[],
+    classes: <ClassShape>[],
+    extensions: <ExtensionShape>[],
+    typedefs: <TypedefShape>[],
+  );
+
+  List<Annotation> summarizeAnnotations(Element element) => [
+        if (element.hasDeprecated) Annotation.deprecated,
+        if (element.hasSealed) Annotation.sealed,
+        if (element.hasVisibleForOverriding) Annotation.visibleForOverriding,
+        if (element.hasVisibleForTesting) Annotation.visibleForTesting,
+      ];
+
+  MethodShape summarizeMethod(MethodElement methodElement) {
+    return MethodShape(
+      name: methodElement.name,
+    );
+  }
+
+  PropertyShape summarizeProperty(PropertyAccessorElement property) {
+    return PropertyShape(name: property.variable.name);
+  }
+
+  NamedConstructorShape summarizeNamedConstructor(
+      ConstructorElement constructor) {
+    return NamedConstructorShape(name: constructor.name);
+  }
+
+  FunctionShape summarizeFunction(FunctionElement functionElement) {
+    return FunctionShape(
+      id: functionElement.id,
+      name: functionElement.name,
+    );
+  }
+
+  GlobalPropertyShape summarizeGlobalProperty(
+      PropertyAccessorElement property) {
+    return GlobalPropertyShape(id: property.id, name: property.variable.name);
+  }
+
+  TypedefShape summarizeTypedef(TypeAliasElement typedefElement) {
+    return TypedefShape(
+      id: typedefElement.id,
+      name: typedefElement.name,
+      targetClassId: typedefElement.aliasedType.element2 is ClassElement
+          ? typedefElement.aliasedType.element2!.id
+          : null,
+    );
+  }
```

Comment:
nit: These could just be top-level private methods, or inline functions expressions where used.

Reply:
made these top-level private methods